### DISCUSSION TITLE
DNS configuration clarification

### DISCUSSION THREAD

**Discussion by dakhnod** [openthread/openthread#10897] [Q&A]
I have observed via `ot-ctl dns config` that child nodes have the leader as dns resolver, and the leader has google as DNS resolver. Regarding this, I have a few questions:

1. Am I right that all children reference the leader as DNS resolver?
2. Under what circumastances does a BR open port 53 for DNS queries?
3. What if the leader is no BR, and thus cannot forward DNS queries from the children to the internet?
4. What minimal config is needed to address mdns hostnames via xxxx.default.service.arpa?

Thanks a lot!

**Comment by abtink** [comment]
OpenThread DNS client is a flexible and stand-alone implementation allowing different configurations based on use-case, including some helpful default behaviors.

Callers can explicitly specify the `otDnsQueryConfig` they want for every query API call (e.g., as input to `otDnsClientResolveAddress()`), which would then be applied to that query.

Callers can also define a "default" `otDnsQueryConfig` which is then used when any configuration properties are not explicitly provided in the resolve/browse API call. Check the documentation of [`otDnsClientSetDefaultConfig()`](https://github.com/openthread/openthread/blob/854e92a40aac4f25db4fc3b103c3b43eb816c9c2/include/openthread/dns_client.h#L137-L162)

You may want to check the feature [`OPENTHREAD_CONFIG_DNS_CLIENT_DEFAULT_SERVER_ADDRESS_AUTO_SET_ENABLE`](https://github.com/openthread/openthread/blob/854e92a40aac4f25db4fc3b103c3b43eb816c9c2/src/core/config/dns_client.h#L86-L101).

**With this feature, the DNS client will discover and track the BRs (which can act as SRP server and DNS resolver) and then auto-select the preferred BR and use its address as the DNS resolver/server address it would use by default in the DNS default `otDnsQueryConfig`.**

Here are answers to your questions:

> Am I right that all children reference the leader as DNS resolver?

No. If `OPENTHREAD_CONFIG_DNS_CLIENT_DEFAULT_SERVER_ADDRESS_AUTO_SET` is enabled (which should be by default), then if there is a BR within the Thread mesh (that is providing SRP server/DNS resolver function), it will be auto-picked.

> Under what circumstances does a BR open port 53 for DNS queries?

All features can be enabled through OT "build config"; e.g., check the `OPENTHREAD_CONFIG_DNSSD_SERVER_ENABLE` feature for this.

If you use `ot-br-posix` to build OTBR, it sets up all configurations properly.

> What if the leader is not a BR, and thus cannot forward DNS queries from the children to the internet?

A BR does not need to be the leader; it can take any role within the Thread mesh. (It often confuses people when we say a BR can even take a child role within the Thread mesh. "Border Router" means it can route between the Thread network and an infra-if network.)
